hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/SparkBaseIndexSupport.scala

```diff
@@ -164,29 +164,34 @@ abstract class SparkBaseIndexSupport(spark: SparkSession,
       var recordKeys: List[String] = List.empty
       for (query <- queryFilters) {
         val recordKeyOpt = getRecordKeyConfig
-        RecordLevelIndexSupport.filterQueryWithRecordKey(query, recordKeyOpt).foreach({
-          case (exp: Expression, recKeys: List[String]) =>
-            recordKeys = recordKeys ++ recKeys
-            recordKeyQueries = recordKeyQueries :+ exp
-        })
+        recordKeyOpt.foreach { recordKeysArray =>
+          // Handle composite record keys
+          RecordLevelIndexSupport.filterQueryWithRecordKey(query, recordKeysArray).foreach {
+            case (exp: Expression, recKeys: List[String]) =>
+              exp match {
+                // For IN, add each element individually to recordKeys
+                case _: In =>
+                  recordKeys = recordKeys ++ recKeys
+
+                // For other cases, basically EqualTo, concatenate recKeys with the default separator
+                case _ =>
+                  recordKeys = recordKeys :+ recKeys.mkString(DEFAULT_RECORD_KEY_PARTS_SEPARATOR)
+              }
+              recordKeyQueries = recordKeyQueries :+ exp
+          }
+        }
       }
-
-      Tuple2.apply(recordKeyQueries, recordKeys)
+      (recordKeyQueries, recordKeys)
     }
   }
 
   /**
-   * Returns the configured record key for the table if it is a simple record key else returns empty option.
+   * Returns the configured record key for the table.
    */
-  private def getRecordKeyConfig: Option[String] = {
+  private def getRecordKeyConfig: Option[Array[String]] = {
     val recordKeysOpt: org.apache.hudi.common.util.Option[Array[String]] = metaClient.getTableConfig.getRecordKeyFields
-    val recordKeyOpt = recordKeysOpt.map[String](JFunction.toJavaFunction[Array[String], String](arr =>
-      if (arr.length == 1) {
-        arr(0)
-      } else {
-        null
-      }))
-    Option.apply(recordKeyOpt.orElse(null))
+    // Convert the Hudi Option to Scala Option and return if present
+    Option(recordKeysOpt.orElse(null)).filter(_.nonEmpty)
```

Comment:
The `nonEmpty` does now work when the record key is null.

Reply:
It does work. Filter would only work on Some option type.